src/jitlayers.cpp

```diff
@@ -190,109 +182,127 @@ void NotifyDebugger(jit_code_entry *JITCodeEntry)
+                auto &Object = *oit;
 #endif
-            auto &LO = *lit;
-
-            OwningBinary<ObjectFile> SavedObject = LO->getObjectForDebug(*Object);
+                auto &LO = *lit;
+
+                OwningBinary<object::ObjectFile> SavedObject = LO->getObjectForDebug(*Object);
+
+                // If the debug object is unavailable, save (a copy of) the original object
+                // for our backtraces
+                if (!SavedObject.getBinary()) {
+                    // This is unfortunate, but there doesn't seem to be a way to take
+                    // ownership of the original buffer
+                    auto NewBuffer = MemoryBuffer::getMemBufferCopy(Object->getData(), Object->getFileName());
+                    auto NewObj = ObjectFile::createObjectFile(NewBuffer->getMemBufferRef());
+                    SavedObject = OwningBinary<object::ObjectFile>(std::move(*NewObj),std::move(NewBuffer));
+                }
+                else {
+                    NotifyGDB(SavedObject);
+                }
 
-            // If the debug object is unavailable, save (a copy of) the original object
-            // for our backtraces
-            if (!SavedObject.getBinary()) {
-                // This is unfortunate, but there doesn't seem to be a way to take
-                // ownership of the original buffer
-                auto NewBuffer = MemoryBuffer::getMemBufferCopy(Object->getData(), Object->getFileName());
-                auto NewObj = ObjectFile::createObjectFile(NewBuffer->getMemBufferRef());
-                SavedObject = OwningBinary<ObjectFile>(std::move(*NewObj),std::move(NewBuffer));
-            }
-            else {
-                NotifyGDB(SavedObject);
+                SavedObjects.push_back(std::move(SavedObject));
+
+                ORCNotifyObjectEmitted(JuliaListener.get(),
+                        *Object,
+                        *SavedObjects.back().getBinary(),
+                        *LO);
+
+                // record all of the exported symbols defined in this object
+                // in the primary hash table for the enclosing JIT
+                for (auto &Symbol : Object->symbols()) {
+                    auto Flags = Symbol.getFlags();
+                    if (Flags & object::BasicSymbolRef::SF_Undefined)
+                        continue;
+                    if (!(Flags & object::BasicSymbolRef::SF_Exported))
+                        continue;
+                    ErrorOr<StringRef> Name = Symbol.getName();
+                    orc::JITSymbol Sym = JIT.CompileLayer.findSymbolIn(H, *Name, true);
+                    assert(Sym);
+                    // note: calling getAddress here eagerly finalizes H
+                    // as an alternative, we could store the JITSymbol instead
+                    // (which would present a lazy-initializer functor interface instead)
+                    JIT.LocalSymbolTable[*Name] = (void*)(uintptr_t)Sym.getAddress();
```

Comment:
IIUC this is the only place we get the symbol address (eagerly) now? It seems that on llvm 39 this is not called before we get to `findSymbol` (maybe because llvm gets lazier?) and therefore `findSymbol` returns `NULL`.